/*
 * Copyright 2018-2020 Qter(qsaker@qq.com). All rights reserved.
 *
 * The file is encoded using "utf8 with bom", it is a part
 * of QtSwissArmyKnife project.
 *
 * QtSwissArmyKnife is licensed according to the terms in
 * the file LICENCE in the root of the source code directory.
 */
#include <QDateTime>

#include "SAKOtherTransmissionPage.hh"
#include "SAKOtherTransmissionItemUdp.hh"
#include "SAKOtherTransmissionItemTcp.hh"
#include "SAKOtherTransmissionItem.hh"
#include "SAKOtherTransmissionItemCom.hh"

#include "ui_SAKOtherTransmissionPage.h"

SAKOtherTransmissionPage::SAKOtherTransmissionPage(SAKDebugPage *debugPage, QWidget *parent)
    :QWidget (parent)
    ,mDebugPage (debugPage)
    ,mUi (new Ui::SAKOtherTransmissionPage)
{
    mUi->setupUi(this);
    mAddItemPushButton = mUi->addItemPushButton;
    mDeleteItemPushButton = mUi->addItemPushButton;
    mListWidget = mUi->listWidget;
    mInfoLabel = mUi->infoLabel;

    mClearMessageInfoTimer.setInterval(SAK_CLEAR_MESSAGE_INTERVAL);
    connect(&mClearMessageInfoTimer, &QTimer::timeout, this, &SAKOtherTransmissionPage::clearMessage);
}

SAKOtherTransmissionPage::~SAKOtherTransmissionPage()
{
    delete mUi;
}

void SAKOtherTransmissionPage::setTransmissionType(TransmissionType type)
{
    mTransmissionType = type;
}

void SAKOtherTransmissionPage::outputMessage(QString msg, bool isInfo)
{
    QString color = "black";
    if (!isInfo){
        color = "red";
        QApplication::beep();
    }
    mInfoLabel->setStyleSheet(QString("QLabel{color:%1}").arg(color));

    mInfoLabel->setText(QTime::currentTime().toString("hh:mm:ss ") + msg);
    mClearMessageInfoTimer.start();
}

void SAKOtherTransmissionPage::clearMessage()
{
    mClearMessageInfoTimer.stop();
    mInfoLabel->clear();
}

void SAKOtherTransmissionPage::on_addItemPushButton_clicked()
{    
    QListWidgetItem *item = new QListWidgetItem(mListWidget);
    mListWidget->insertItem(mListWidget->count(), item);
    QWidget *itemWidget = Q_NULLPTR;
    switch (mTransmissionType){
    case SerialPortTransmission:
        itemWidget = new SAKOtherTransmissionItemCom(mDebugPage, this);
        break;
    case UdpTransmission:
        itemWidget = new SAKOtherTransmissionItemUdp(mDebugPage, this);
        break;
    case TcpTransmission:
        itemWidget = new SAKOtherTransmissionItemTcp(mDebugPage, this);
        break;
    default:
        Q_ASSERT_X(false, __FUNCTION__, "Unknown transmissioin type");
        break;
    }
    item->setSizeHint(QSize(itemWidget->width(), itemWidget->height()));
    mListWidget->setItemWidget(item, itemWidget);
    SAKOtherTransmissionItem *baseItemWidget = reinterpret_cast<SAKOtherTransmissionItem*>(itemWidget);
    connect(baseItemWidget, &SAKOtherTransmissionItem::requestOutputMessage, this, &SAKOtherTransmissionPage::outputMessage);
}

void SAKOtherTransmissionPage::on_deleteItemPushButton_clicked()
{
    QListWidgetItem *item = mListWidget->currentItem();
    mListWidget->removeItemWidget(item);
    delete item;
}
